derive-c-marshalling/src/lib.rs

```diff
@@ -0,0 +1,167 @@
+#![recursion_limit="128"]
+
+extern crate libc;
+extern crate proc_macro;
+extern crate syn;
+#[macro_use]
+extern crate quote;
+
+use proc_macro::TokenStream;
+
+#[proc_macro_derive(CMarshalling)]
+pub fn hello_world(input: TokenStream) -> TokenStream {
+
+    let ast = syn::parse_derive_input(&input.to_string()).unwrap();
+
+    let cstring_type = ::syn::parse_type(stringify!(::std::ffi::CString)).unwrap();
+    // TODO(gardell): Combine absolute and relative type checking by expanding the compared type.
```

Comment:
Rather than comparing string type found in the AST it would be more robust to add and implement a trait which does conversion between C types and Rust types.

It is additional work though so maybe something to work towards long term (as this works for the use case).